backend/src/algorithm/index.ts

```diff
@@ -0,0 +1,60 @@
+import { daysPerWeek, maxPeriodsPerDay } from './consts';
+import { GenerateTimetableProps, GenerateTimetableResult, RemainingLectures } from './types';
+import { getTotalRemainingLectures, getTotalTimesPerWeek } from './utils/subject';
+import { setSlot, swapSlots } from './utils/slot';
+import { checkConstraints } from './utils/constraint';
+import { generateTimetableProps } from './seed';
+import { getClassPeriodsPerDay } from './utils/class';
+import { getMatrixHashmap } from './utils/matrix';
+import { shuffleArray } from './utils/array';
+
+export function generateTimetable({ classes, classrooms }: GenerateTimetableProps): GenerateTimetableResult {
+  const teachers = classes
+    .map(class_ => class_.subjects.map(subject => subject.teacher))
+    .flat()
+    .filter((teacher, i, arr) => arr.findIndex(ts => ts.name === teacher.name) === i);
+
+  const timetable = getMatrixHashmap(classes.map(class_ => class_.name));
+  // holds current state for every teacher and classroom occupancy at certain time slot
+  const unavailable = {
+    teachers: getMatrixHashmap(teachers),
+    classrooms: getMatrixHashmap(classrooms),
+  };
+
+  // holds current state for every subject lecture count
+  const remainingLectures: RemainingLectures = Object.fromEntries(
+    classes.map(class_ => [
+      class_.name,
+      Object.fromEntries(class_.subjects.map(subject => [subject.name, getTotalTimesPerWeek(subject)])),
+    ]),
+  );
+  // holds daily total period quantity for every class
+  const periods = Object.fromEntries(classes.map(class_ => [class_.name, getClassPeriodsPerDay(class_)]));
+
+  for (let dayIndex = 0; dayIndex < daysPerWeek; ++dayIndex) {
+    for (let periodIndex = 0; periodIndex < maxPeriodsPerDay; ++periodIndex) {
+      // shuffle classes for for more equal classroom distribution
```

Comment:
Lint: double for word in comment.

Reply:
It is still there 😄

Reply:
Sneaky sneaky 😂 Was sad while removing it, but it had to be done. 😁